sklearn/impute/_knn.py

```diff
@@ -217,71 +217,81 @@ def transform(self, X):
-                X[all_nan_receivers_idx, col] = col_mean
+                potential_donors_idx, = np.nonzero(non_missing_fix_X[:, col])
 
-                if len(all_nan_receivers_idx) == len(receivers_idx):
-                    # all receivers imputed with mean
-                    continue
+                # receivers_idx are indices in X
+                receivers_idx = row_missing_chunk[np.flatnonzero(col_mask)]
 
-                # receivers with at least one defined distance
-                receivers_idx = receivers_idx[~all_nan_dist_mask]
-                dist_subset = (dist[dist_idx_map[receivers_idx]]
+                # distances for samples that needed imputation for column
+                dist_subset = (dist_chunk[dist_idx_map[receivers_idx] - start]
                                [:, potential_donors_idx])
 
-            n_neighbors = min(self.n_neighbors, len(potential_donors_idx))
-            value = self._calc_impute(dist_subset, n_neighbors,
-                                      self._fit_X[potential_donors_idx, col],
-                                      mask_fit_X[potential_donors_idx, col])
-            X[receivers_idx, col] = value
-
-        return super()._concatenate_indicator(X[:, valid_idx], X_indicator)
+                # receivers with all nan distances impute with mean
+                all_nan_dist_mask = np.isnan(dist_subset).all(axis=1)
+                all_nan_receivers_idx = receivers_idx[all_nan_dist_mask]
+
+                if all_nan_receivers_idx.size:
+                    col_mean = np.ma.array(self._fit_X[:, col],
+                                           mask=mask_fit_X[:, col]).mean()
+                    X[all_nan_receivers_idx, col] = col_mean
+
+                    if len(all_nan_receivers_idx) == len(receivers_idx):
+                        # all receivers imputed with mean
+                        continue
+
+                    # receivers with at least one defined distance
+                    receivers_idx = receivers_idx[~all_nan_dist_mask]
+                    dist_subset = (dist_chunk[dist_idx_map[receivers_idx]
+                                              - start]
+                                   [:, potential_donors_idx])
+
+                n_neighbors = min(self.n_neighbors, len(potential_donors_idx))
+                value = self._calc_impute(
+                    dist_subset,
+                    n_neighbors,
+                    self._fit_X[potential_donors_idx, col],
+                    mask_fit_X[potential_donors_idx, col])
+                X[receivers_idx, col] = value
+
+        # process in fixed-memory chunks
+        gen = pairwise_distances_chunked(
+            X[row_missing_idx, :],
+            self._fit_X,
+            metric=self.metric,
+            missing_values=self.missing_values,
+            force_all_finite=force_all_finite,
+            reduce_func=process_chunk)
+        for chunk in gen:
```

Comment:
do we need this for loop?

Reply:
Do you have another proposal. One recipe to consume generator is to use a Python deque

```python
from collections import deque
deque(gen, maxlen=0)
```

but I am not sure this is more intuitive for most people.

Reply:
We can think of all the fun ways to consume the generator (without using its output). :)

I think the for loop is the most straight forward.

Reply:
We could a different manner each time we can :)

Reply:
it's just to consume the iterable? ok